latest/index.bs

```diff
@@ -175,7 +175,292 @@ For this example we assume an image with 5 dimensions and axes called `t,c,z,y,x
+        |   |
+        |   ├── .zattrs       # `.zattrs` MUST contain `"_index"`, which is the name of the column in obs to be used as the index.           
+        |   │                 # `.zattrs` MUST contain `"column-order"`, which is a list of the order of the non-_index columns.
+        |   │                 # `.zattrs` MUST contain `"encoding-type"`, which is set to `"dataframe"` by AnnData.
+        |   │                 # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.2.0"` by AnnData.
+        |   │
+        |   ├── array_col     # Columns in the var table MAY be a 1D zarr array. The rows can be chunked as the user desires.
+        |   |   ├── .zarray   # However, the var columns SHOULD be chunked in the same way as the columns in X.
+        |   |   │
+        |   |   └─ 0
+        |   |
+        |   └── cat_col       # Columns in the var table MAY be categorical
+        |       ├── .zattrs.  # `.zattrs` MUST contain `"encoding-type"`, which is set to `"categorical"` by AnnData.
+        |       |             # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.2.0"` by AnnData.
+        |       |
+        |       ├── categories
+        |       |  ├── .zarray # categories MUST be a 1D zarr array. The rows can be chunked as the user desires.
+        |       |  |
+        |       |  └─ 0
+        |       ├── codes
+        |       |  ├── .zarray # codes MUST be a 1D zarr array. The rows can be chunked as the user desires.
+        |       |  |
+        |       |  └─ 0
+        |       |
+        |       ├── null_col  # Columns in the var table MAY nullable integer
+        |       ├── .zattrs.  # `.zattrs` MUST contain `"encoding-type"`, which is set to `"nullable-integer"` by AnnData.
+        |       |             # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.1.0"` by AnnData.
+        |       |
+        |       ├── mask
+        |       |  ├── .zarray # categories MUST be a 1D zarr array. The rows can be chunked as the user desires.
+        |       |  |
+        |       |  └─ 0
+        |       └── values
+        |          ├── .zarray # codes MUST be a 1D zarr array. The rows can be chunked as the user desires.
+        |          |
+        |          └─ 0
+        |
+        ├── obsm              # You MAY add a obsm group comtainer. The obsm group contains arrays that annotate the rows in X.
+        |   │                 # The rows in each array MUST be index-matched to the rows in X (if present). 
+        |   |
+        │   ├── .zgroup
+        |   |
+        |   ├── .zattrs       # `.zattrs` MUST contain `"encoding-type"`, which is set to `"dict"` by AnnData.           
+        |   │                 # `.zattrs` MUST contain `"encoding-version"`, which is set to `"0.1.0"` by AnnData.
+        |   │
+        │   └── obsm_0        # You MAY add a zarr array for each obsm matrix.
```

Comment:
I don't see where `obsm_0` is referenced in the parent `.zattrs`?
How do we know to load the `obsm_0` data (if we can't `ls` the directories below `obsm`)?
The `obs/.zattrs` has `"column-order"` which lists columns, but there isn't an equivalent for `obsm/.zattrs`.

Reply:
Same issue exists for `obsp`, `varm` and `varp`?

Reply:
We've basically assumed that you can `ls`, and not maintained order for these.

Reply:
Unfortunately that's not the case if you're loading data over `https` (or for some `s3` backends), so we've avoided making that assumption elsewhere on the spec to allow web-based accessing of OME-NGFF.

Reply:
:+1: for the locations of everything in the hierarchy being either computable (implementations know how to figure it out) or explicitly listed in the metadata.

FWIW, the OME2022 zarr-java discussion today touched on exactly this point and the need to perhaps bubble this requirement up to the zarr spec itself.

Reply:
Isn't this solved by consolidated metadata?

Reply:
So, the spec would require that the `image.zarr/table/.zmetadata` MUST exist and contain the consolidated metadata?
Or would we want the metadata consolidated at a different level of the hierarchy?

That is probably the easiest solution, since it's easy to add those steps onto the table creation (at least in python):

```
    store = zarr.DirectoryStore("test_segment.zarr/table")
    zarr.consolidate_metadata(store)
```

Reply:
Added my opinion to https://github.com/kevinyamauchi/ome-ngff-tables-prototype/pull/12 but I'll reiterate here: consolidate_metadata isn't in the v2 spec and isn't supported by all implementations. I would not suggest we go that route unless absolutely necessary.

Reply:
> So, the spec would require that the image.zarr/table/.zmetadata MUST exist and contain the consolidated metadata?
> Or would we want the metadata consolidated at a different level of the hierarchy?

My thinking was more along the lines of:

"If you are providing-zarrs-via/ writing-zarrs-to stores where `ls` is slow or unavailable – like some cloud backends – you should be using consolidated metadata."

So, not a hard requirement, but more of a recommendation about how to configure your zarr store for this circumstance.

Reply:
But there's no way for that to be discovered. We'd essentially need to work into the spec "you must always try with and without (or vice versa) consolidated metadata".

Reply:
Let's consider the 2 issues separately, since I think the cost/benefit of consolidated metadata are different...

For the top-level `image.zarr/tables/.zattrs` it is very easy to manually add the metadata to list the child tables as in https://github.com/kevinyamauchi/ome-ngff-tables-prototype/pull/12. This should be a requirement (MUST) in the spec so all readers can rely on it (without needing to try consolidated metadata).

However, in the other case, I don't know how easy it is to adopt a similar approach for the `obsm, obsp, layers, raw and uns` groups? I'm not very familiar with AnnData and haven't dug into the creation of these groups.

Using a naive approach, I tried simply listing the sub-directories of these groups, and adding those names to the `.zattrs` of each group. 

E.g. add this to `write_table_regions()` from the `ome-ngff-tables-prototype`:

```
    # write group names into .zattrs
    table_path = os.path.join(group.store.dir_path(), group.path, table_group_name)
    for sub_dir in ["layers", "obsm", "obsp", "raw", "uns"]:
        sub_path = os.path.join(table_path, sub_dir)
        if not os.path.exists(sub_path):
            continue
        children = [f.name for f in os.scandir(sub_path) if f.is_dir()]
        sub_group = table_group[sub_dir]
        sub_group.attrs[sub_dir] = children
```

E.g. `obsm/.zattrs` now looks like this:

```
{
    "encoding-type": "dict",
    "encoding-version": "0.1.0",
    "obsm": [
        "X_scanorama",
        "X_umap",
        "spatial"
    ]
}
```

This approach feels a bit hacky, and would need to be recursive in some cases (e.g. uns).
I don't have a strong preference for this compared with `consolidate_metadata()`.
These sets of metadata aren't the core matrix data for the ann-data table. They're not even displayed in napari (in the screenshot) so it's less critical than listing the tables above.

If a client relies on `consolidate_metadata()` and it's not available, then they simply wouldn't display this metadata.
A client wouldn't need to always try with and without consolidated metadata.

Reply:
> If a client relies on consolidate_metadata() and it's not available, then they simply wouldn't display this metadata.

Having data disappear because of something that did or did not get called in the python code and is otherwise not recorded doesn't feel great.

Reply:
@joshmoore So you prefer the `# write group names into .zattrs` approach above (like a manual consolidate metadata)? Or another alternative?

The presence or absence of everything in the viewers is always depends on whether the creating code did or did not add something. That's the choice that the code has when the spec says `SHOULD` or `MAY`.

Reply:
All other things being equal, I prefer that all details & members of a fileset are _either_ well-known beforehand (i.e. static in the spec) _or_ can be determined from the metadata.